dialogue-annotations-processor/src/main/java/com/palantir/dialogue/annotations/processor/generate/ServiceImplementationGenerator.java

```diff
@@ -166,21 +167,34 @@ private Optional<FieldSpec> deserializer(ReturnType type) {
         ParameterizedTypeName deserializerType =
                 ParameterizedTypeName.get(ClassName.get(Deserializer.class), innerType);
 
-        CodeBlock realDeserializer = CodeBlock.of(
+        CodeBlock deserializer = CodeBlock.of(
                 "new $T<>(new $T(), new $T()).deserializerFor(new $T<$T>() {})",
                 ErrorHandlingDeserializerFactory.class,
                 deserializerFactoryType,
                 errorDecoderType,
                 TypeMarker.class,
                 innerType);
-        CodeBlock voidDeserializer = CodeBlock.of(
-                "new $T($L.bodySerDe().emptyBodyDeserializer(), new $T())",
-                ErrorHandlingVoidDeserializer.class,
-                serviceDefinition.conjureRuntimeArgName(),
-                errorDecoderType);
+
+        // If no custom deserializer is used, use the runtime-provided deserializers for specific types
+        if (!type.isUsingCustomDeserializer()) {
+            if (type.isVoid()) {
+                deserializer = CodeBlock.of(
+                        "new $T($L.bodySerDe().emptyBodyDeserializer(), new $T())",
+                        ErrorHandlingVoidDeserializer.class,
+                        serviceDefinition.conjureRuntimeArgName(),
+                        errorDecoderType);
+            } else if (type.isInputStream()) {
```

Comment:
should we handle `Optional<InputStream>` as well? https://github.com/palantir/dialogue/blob/85bb924ae5773397bb8248198d7a0e3c943b3946/dialogue-target/src/main/java/com/palantir/dialogue/BodySerDe.java#L44-L45